observationtree/src/main/scala/explore/observationtree/ConstraintSetObsList.scala

```diff
@@ -0,0 +1,383 @@
+      constraintSetsWithObs: View[ConstraintSetsWithObs],
+      obsId:                 Observation.Id,
+      obsWithIndexGetAdjust: GetAdjust[ObsList, obsListMod.ElemWithIndex]
+    )(implicit
+      c:                     TransactionalClient[IO, ObservationDB]
+    ): Option[Option[ConstraintSet.Id]] => IO[Unit] = { csIdOpt =>
+      val obsCsAdjuster = obsWithIndexGetAdjust
+        .composeOptionLens(first)
+        .composeOptionLens(ObsSummary.constraintSetId)
+
+      val observationsView = constraintSetsWithObs.zoom(ConstraintSetsWithObs.obs)
+
+      // 1) Update internal model
+      observationsView.mod(obsCsAdjuster.set(csIdOpt)) >>
+        // 2) Send mutation
+        csIdOpt.map(newId => moveObs(obsId, newId)).orEmpty
+
+    }
+
+    protected def onDragEnd(
+      setter:      Undoer.Setter[IO, ConstraintSetsWithObs],
+      expandedIds: View[SortedSet[ConstraintSet.Id]]
+    )(implicit
+      c:           TransactionalClient[IO, ObservationDB]
+    ): (DropResult, ResponderProvided) => IO[Unit] =
+      (result, _) =>
+        $.propsIn[IO] >>= { props =>
+          result.destination.toOption
+            .map(destination =>
+              (Observation.Id.parse(result.draggableId) match {
+                case Some(obsId) =>
+                  val obsWithId: GetAdjust[ObsList, obsListMod.ElemWithIndex] =
+                    obsListMod.withKey(obsId)
+
+                  val set: Option[Option[ConstraintSet.Id]] => IO[Unit] =
+                    setter.set[Option[Option[ConstraintSet.Id]]](
+                      props.constraintSetsWithObs.get,
+                      getConstraintSetForObsWithId(obsWithId.getter).get,
+                      setConstraintSetForObsWithId(props.constraintSetsWithObs, obsId, obsWithId)
+                    )
+
+                  destination.droppableId match {
+                    case UnassignedObsId           => set(none.some)
+                    case ConstraintSet.Id(newCsId) =>
+                      expandedIds.mod(_ + newCsId) >> set(newCsId.some.some)
+                    case _                         => IO.unit
+                  }
+                case None        => IO.unit
+              })
+            )
+            .orEmpty
+        }
+
+    def toggleExpanded(
+      id:          ConstraintSet.Id,
+      expandedIds: View[SortedSet[ConstraintSet.Id]]
+    ): Callback =
+      expandedIds
+        .mod(expanded => expanded.exists(_ === id).fold(expanded - id, expanded + id))
+        .runAsyncCB
+
+    // Adapted from https://github.com/atlassian/react-beautiful-dnd/issues/374#issuecomment-569817782
+    def getDraggedStyle(style:   TagMod, snapshot: Draggable.StateSnapshot): TagMod =
+      if (!snapshot.isDragging)
+        TagMod.empty
+      else if (!snapshot.isDropAnimating)
+        style
+      else
+        TagMod(style, ^.transitionDuration := "0.001s")
+
+    def getListStyle(isDragging: Boolean): TagMod =
+      ExploreStyles.DraggingOver.when(isDragging)
```

Comment:
At some point we probably want to abstract away common logic with other views. We can do that later, but can we do it with these 2 methods (and maybe `renderObsBadge`) where the code is the same as in `TargetObsList`?

Reply:
👍